mlflow/data/numpy_dataset.py

```diff
@@ -0,0 +1,156 @@
+            flattened_targets = self._targets.flatten()
+            trimmed_targets = flattened_targets[0:MAX_ROWS]
+            try:
+                md5.update(pd.util.hash_array(trimmed_targets))
+            except TypeError:
+                md5.update(np.int64(trimmed_targets.size))
+            # hash full feature dimensions
+            for x in self._targets.shape:
+                md5.update(np.int64(x))
+
+        return md5.hexdigest()[:8]
+
+    def _to_dict(self, base_dict: Dict[str, str]) -> Dict[str, str]:
+        """
+        :param base_dict: A string dictionary of base information about the
+                          dataset, including: name, digest, source, and source
+                          type.
+        :return: A string dictionary containing the following fields: name,
+                 digest, source, source type, schema (optional), profile
+                 (optional).
+        """
+        base_dict.update(
+            {
+                "schema": json.dumps({"mlflow_tensorspec": self.schema.to_dict()}),
+                "profile": json.dumps(self.profile),
+            }
+        )
+        return base_dict
+
+    @property
+    def source(self) -> FileSystemDatasetSource:
+        return self._source
+
+    @property
+    def features(self) -> Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]]:
+        return self._features
+
+    @property
+    def targets(self) -> Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]]:
+        return self._targets
+
+    @property
+    def profile(self) -> Optional[Any]:
+        """
+        A profile of the dataset. May be None if no profile is available.
+        """
+        return {
+            "shape": self._features.shape,
+        }
+
+    @cached_property
+    def schema(self) -> Schema:
+        """
+        An MLflow TensorSpec schema representing the tensor dataset
+        """
+        return _infer_schema(self._features)
+
+    def to_pyfunc(self) -> PyFuncInputsOutputs:
+        """
+        Converts the dataset to a collection of pyfunc inputs and outputs for model
+        evaluation. Required for use with mlflow.evaluate().
+        May not be implemented by all datasets.
+        """
+        return PyFuncInputsOutputs(self._features, self._targets)
+
+
+def from_numpy(
+    features: Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]],
+    source: Union[str, DatasetSource],
+    targets: Union[np.ndarray, List[np.ndarray], Dict[str, np.ndarray]] = None,
+    name: Optional[str] = None,
+    digest: Optional[str] = None,
+) -> NumpyDataset:
+    """
+    :param features: NumPy features, represented as an np.ndarray, list of np.ndarrays
+                    or dictionary of named np.ndarrays.
+    :param source: The source from which the NumPy data was derived, e.g. a filesystem
+                    path, an S3 URI, an HTTPS URL etc. Attempting to use other source
+                     types will throw.
```

Comment:
> Attempting to use other source types will throw

It won't throw if the source is resolvable, right? Can we remove this and also include some info about specifying a `DatasetSource` object?